File: goal解析器.cpp
#include <stdio.h>
#include <string>
#include <iostream>

using namespace std;


enum Status {
  normal,
  first_bracket,
  in_bracket
};


class Solution {
  public:
    string interpret(string command) {
      string result;
      Status status;
      for (auto i = command.begin(); i != command.end(); i++) {
        if (*i == 'G') {
          result += 'G';
          continue;
        } else if (*i == '(') {
          status = first_bracket;
          continue;
        } else if (*i == ')') {
          if (status == in_bracket) {
            result += "al";
          } else if (status == first_bracket) {
            result += 'o';
          }
          status = normal;
          continue;
        } else if (*i == 'a') {
          status = in_bracket;
          continue;
        }
      }
      return result;
    }
};

int main() {
  string input = "G()()()(al)";
  string output = Solution().interpret(input);
  cout << output << endl;
};

File: goal解析器.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Xiang Wang <ramwin@qq.com>


import enum


class Status(enum.Enum):
    NORMAL = 1
    FIRST_BRACKET = 2
    IN_BRACKET = 3


class Solution:

    def interpret(self, command: str) -> str:
        result = ""
        status = Status.NORMAL
        for i in command:
            if i == "G":
                result += "G"
            elif i == "(":
                status = Status.FIRST_BRACKET
            elif i == ")":
                if status == Status.FIRST_BRACKET:
                    result += "o"
                elif (status == Status.IN_BRACKET):
                    result += 'al'
            else:
                status = Status.IN_BRACKET
        return result


print(Solution().interpret("G()(al)"))
